rcl/test/rcl/test_publisher.cpp

```diff
@@ -290,3 +290,94 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_init_
 }
+
+TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_loan) {
+  rcl_publisher_t publisher = rcl_get_zero_initialized_publisher();
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, Strings);
+  const char * topic_name = "chatter";
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  rcl_ret_t ret =
+    rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+
+  test_msgs__msg__Strings * msg_loaned = nullptr;
+  test_msgs__msg__Strings ** msg_loaned_ptr = &msg_loaned;
+  if (rcl_publisher_can_loan_messages(&publisher)) {
+    EXPECT_EQ(
+      RCL_RET_OK, rcl_borrow_loaned_message(
+        &publisher,
+        ts,
+        reinterpret_cast<void **>(msg_loaned_ptr)));
+    ASSERT_TRUE(rosidl_runtime_c__String__assign(&(msg_loaned->string_value), "testing"));
+    EXPECT_EQ(
+      RCL_RET_OK, rcl_publish_loaned_message(
+        &publisher,
+        msg_loaned,
+        nullptr));
+  }
+}
+
+TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_publisher_access_functions) {
+  rcl_publisher_t publisher = rcl_get_zero_initialized_publisher();
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, Strings);
+  const char * topic_name = "chatter";
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  rcl_ret_t ret =
+    rcl_publisher_init(&publisher, this->node_ptr, ts, topic_name, &publisher_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_publisher_fini(&publisher, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+
+  const rcl_publisher_options_t * publisher_options_rcv = rcl_publisher_get_options(&publisher);
+  ASSERT_NE(nullptr, publisher_options_rcv);
+  EXPECT_EQ(rmw_qos_profile_default.reliability, publisher_options_rcv->qos.reliability);
+  EXPECT_EQ(rmw_qos_profile_default.history, publisher_options_rcv->qos.history);
+  EXPECT_EQ(rmw_qos_profile_default.depth, publisher_options_rcv->qos.depth);
+  EXPECT_EQ(rmw_qos_profile_default.durability, publisher_options_rcv->qos.durability);
+  EXPECT_TRUE(rcutils_allocator_is_valid(&(publisher_options_rcv->allocator)));
+
+  rmw_publisher_t * pub_rmw_handle = rcl_publisher_get_rmw_handle(&publisher);
+  EXPECT_NE(nullptr, pub_rmw_handle);
+
+  rcl_context_t * pub_context = rcl_publisher_get_context(&publisher);
+  EXPECT_TRUE(rcl_context_is_valid(pub_context));
+  EXPECT_EQ(rcl_context_get_instance_id(context_ptr), rcl_context_get_instance_id(pub_context));
+
+  EXPECT_EQ(nullptr, rcl_publisher_get_options(nullptr));
```

Comment:
I think it would be good make sure that all cases checked by `rcl_publisher_is_valid` result in the expected output in these functions. Besides nullptr, that would be `publisher->impl` is null and `publisher->impl->rmw_handle` is null, and publisher->impl->context is null.

Reply:
Given that all of these functions use internally `rcl_publisher_is_valid` or `rcl_publisher_is_valid_except_context` I think the best idea would be adding tests in particular for those two functions.

However, I think the second case you mention can't be achieved with use of the public API/definitions. Should I add the private part of the implementation to make that test?